libs-scala/contextualized-logging/src/main/scala/com/digitalasset/logging/LoggingValue.scala

```diff
@@ -0,0 +1,63 @@
+// Copyright (c) 2021 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.logging
+
+import java.time.{Duration, Instant}
+
+import scala.collection.SeqView
+import scala.language.implicitConversions
+
+sealed trait LoggingValue
+
+object LoggingValue {
+  final object Empty extends LoggingValue
+
+  final case class OfString(value: String) extends LoggingValue
+
+  final case class OfInt(value: Int) extends LoggingValue
+
+  final case class OfLong(value: Long) extends LoggingValue
+
+  final case class OfIterable(sequence: Iterable[LoggingValue]) extends LoggingValue
```

Comment:
You _could_ avoid the inner boxing layer on the elements by adding an existential here. It's always the same for every element, anyway. Then you'd add some GADT stuff to `LoggingValueSerializer`...totally up to you whether you want to try to get that working, it's a not-insignificant design change.

Reply:
I _think_ I understand what you mean here. I was trying to avoid type parameters at all for `LoggingValue` (hence `ToLoggingValue`'s existence), mostly because I think it will pollute the API. Can you see a way to do this without making the API more confusing?

Reply:
Type parameters occur only within the data constructors; I do not consider `OfIterable` to be a part of the API, so to speak.

Reply:
Oh, I see, that might work. I'll look into it but in a separate change; as you said, it's not insignificant and I'm not so knowledgeable about existentials.

If you feel like doing it at some point today, feel free. Otherwise I'll give it a crack once this is in.